Improved Object Validations with New ValidateObject

Abhishek-Punhani opened the pull request:
Implemented improved object validations using validateObject function

- Replaced ad-hoc validation logic with validateObject function for consistency and maintainability
- Updated prop validation in components to use validateObject
- Ensure all object prop validations are thorough and standardized
- Manual verification performed to ensure no regressions

Issue Addressed : #8903

Abhishek-Punhani committed:
App: Learn

Improved Object Validations with New ValidateObject

MisRob commented:
Thanks @Abhishek-Punhani! I skimmed briefly and overall looks as what I would expect. I will invite my colleague @AllanOXDi for more detailed reviews.

AllanOXDi commented on kolibri/plugins/learn/assets/src/views/BrowseResourceMetadata.vue:
```diff
@@ -265,6 +266,25 @@
+            description: { type: String, required: false },
+            duration: { type: Number, required: false },
+            grade_levels: { type: Array, required: false },
+            lang: { type: Object, required: false },
```
I think `lang ` store a language code ( "en" or "es"), it should be a String? @MisRob  What do you think here?

Abhishek-Punhani replied:
Hi, @AllanOXDi  In the same file, at line 94, I noticed that content.lang.lang_name is being accessed, which is why I defined lang as an object. Thanks for pointing out this; please let me know your suggestion on this.

MisRob replied:
Thanks both. In the screenshot I posted below `lang` is one of the console errors, so let's see if it's resolved after the PR is updated with fixes.

AllanOXDi commented on kolibri/plugins/learn/assets/src/views/CardList.vue:
```diff
@@ -157,6 +158,28 @@
+            description: { type: String, required: false },
+            duration: { type: Number, required: false },
+            grade_levels: { type: Array, required: false },
+            lang: { type: Object, required: false },
```
The lang object also needs a default value that creates a new object each time. It should be default: () => ({})

Abhishek-Punhani replied:
I have implemented the requested changes by setting the lang object's default value to default: () => ({}). Let me know if any further modifications are needed.

AllanOXDi reviewed (commented):
This looks good overall, but I have a couple of suggestions.

Abhishek-Punhani committed:
Added a default value to lang in Cardlist.vue

MisRob commented on kolibri/plugins/learn/assets/src/views/AlsoInThis.vue:
```diff
@@ -148,18 +148,31 @@
         type: Array,
         required: true,
         default: () => [],
+        /** Content node with the following properties: id, is_leaf, title */
```
I think we could even remove comments like this. Nice thing about `validateObject` is that it is self-documenting.

MisRob reviewed (changes requested):
Hi @Abhishek-Punhani, I see many errors in the console that appear to be caused by the updates in this pull request. They are not present in the `develop` branch.

![Screenshot from 2025-02-25 23-11-25](https://github.com/user-attachments/assets/49f8fb69-7a7a-4702-afbc-34aa5b32ca04)

The goal here is to implement validations in a way that fits exactly how the objects are used, or fix problematic usages that fail the validations.

This is quite a lot of cumbersome work that requires thorough examination of code as well as manual testing, so feel free to go slowly step by step - there is really no pressure around amounts of work and such.

Abhishek-Punhani committed:
Fixed Console Errors

Corrected Some Incorrect Type Assignments

Abhishek-Punhani commented:
@MisRob , There were some incorrect type assignments; I think now all errors are resolved; please check!

AllanOXDi reviewed (approved):
looks good to me.  Thanks @Abhishek-Punhani  - I will let my colleague pass a second eye to it

MisRob commented:
Thank you, I will have a look. @Abhishek-Punhani we are now in the period of time shortly before release, so I wanted to note even if all will look good, it may take a while to merge - I will take some time to coordinate with relevant people.

MisRob reviewed (changes requested):
Hi @Abhishek-Punhani and @AllanOXDi,

there are still issues, for example here (there may be more)

![Screenshot from 2025-03-05 20-52-32](https://github.com/user-attachments/assets/3d8d11c1-d2fa-4fe8-ad9c-903a6dedd132)

During development and review, I think it'd be best use Vue Devtools + examining imports in code to find what places need to be previewed, and carefully test step by step.

@Abhishek-Punhani Some components are used at more places, and this task is a bit more tricky than I expected - you may consider decreasing the size of this PR to move this work on smoothly. Perhaps just take care of one or two components, but find a way to test it everywhere where it's used.

Abhishek-Punhani commented:
@MisRob , I am unable to reproduce these console errors. Could you please guide me in doing so?

Abhishek-Punhani commented:
@MisRob , I didn't import all resources earlier; I saw errors on the console; I will look into them and make changes asap.

Abhishek-Punhani commented:
@MisRob , I don't see errors now, but there's some sort of confusion around accessibility_labels and grade_levels regarding whether these should be assigned Object or Array. I have made some changes and don't see any errors now; please check at your convenience!

MisRob commented:
Thanks for the updates @Abhishek-Punhani. @AllanOXDi would you be available to re-review?

Before we proceed with the new review, @Abhishek-Punhani I'd like to confirm if you were able to test this PR fully as [suggested here](https://github.com/learningequality/kolibri/pull/13081#pullrequestreview-2662352995) or do we need to consider decreasing the scope?

MisRob commented:
> I didn't import all resources earlier; I saw errors on the console; I will look into them and make changes asap.

Ah yes that could be the cause, thanks for looking into it before I could reply. Well, we will try as best as we can to take care of testing as many places we can, even though we may miss some contexts. I am quite not sure yet how to best organize around these kinds of changes, let's keep trying :)

I also plan to invite QA team later, however we will need to provide them with information about what pages to preview since they won't read this from the code. Perhaps @Abhishek-Punhani in your upcoming PRs, you may jot down brief note with the list of places that were affected and then we can all use it to sync?

Abhishek-Punhani commented:
> Thanks for the updates @Abhishek-Punhani. @AllanOXDi would you be available to re-review?
> 
> Before we proceed with the new review, @Abhishek-Punhani I'd like to confirm if you were able to test this PR fully as [suggested here](https://github.com/learningequality/kolibri/pull/13081#pullrequestreview-2662352995) or do we need to consider decreasing the scope?

Yes, @MisRob, I saw all those errors and fixed them; I was confused over accessibility labels and grade levels. However, assigning them type Array and validating them as both Array and Object fixes the errors. Some errors were due to my previous PR for the exact cause of type assignment to  accessibility_labels and grade_levels. I have fixed them too. While reviewing, if you see new errors, please let me know; thanks!

Abhishek-Punhani commented:
> > I didn't import all resources earlier; I saw errors on the console; I will look into them and make changes asap.
> 
> Ah yes that could be the cause, thanks for looking into it before I could reply. Well, we will try as best as we can to take care of testing as many places we can, even though we may miss some contexts. I am quite not sure yet how to best organize around these kinds of changes, let's keep trying :)
> 
> I also plan to invite QA team later, however we will need to provide them with information about what pages to preview since they won't read this from the code. Perhaps @Abhishek-Punhani in your upcoming PRs, you may jot down brief note with the list of places that were affected and then we can all use it to sync?

Yes, @MisRob , It sounds like a good idea; I’ll make sure to include a brief note in my upcoming PRs about the affected areas to help with syncing.

MisRob commented:
Thank you @Abhishek-Punhani.

I will let @AllanOXDi re-review. Allan during the review, would you please try to confirm there are no unexpected errors at the places where these components are used? I can give it a look too later but more eyes here will be needed. Thank you.

AllanOXDi reviewed (changes requested):
Hi @Abhishek-Punhani  Thanks for the great work here . I think these changes are till causing regressions and I can still see errors @MisRob reported in the console and in few other places. See the screen grab bellow. 
<img width="463" alt="Screenshot 2025-03-18 at 11 54 17" src="https://github.com/user-attachments/assets/46884276-49f9-43e6-8ec6-dba1aea95206" />
<img width="935" alt="Screenshot 2025-03-18 at 11 47 22" src="https://github.com/user-attachments/assets/8a7df526-31dd-481f-9139-fc93c453356c" />

MisRob commented:
Thank you @AllanOXDi for thorough testing. @Abhishek-Punhani I think this would be a good time to decrease size of this pull request as I suggested before to make sure we can finish it. Would you be able to choose 1-3 components, definitely not more than 3, and keep only those changes in this pull request, while re-testing them one more time? Also if you'd include the list of affected places, that will help us with re-reviews.

Abhishek-Punhani commented:
Hi @AllanOXDi , Can you please help me reproduce those console errors? I don't see any errors on my local host.
@MisRob So, should I close this PR and raise new PRs with fewer files changed, or should we do so from the next PR because we have reviewed these changes a lot? Please let me know what you think!

MisRob commented:
> So, should I close this PR and raise new PRs with fewer files changed

Yes @Abhishek-Punhani, let's do it right now. Just choose maximum of 3 components. I'm sorry we will loose some of that other work, but you can keep it in a separate branch for future and gradually pull out from there.

In a new PR, you can also include that list with places to test and then we can all have a look at it one more time and figure out better for each other steps to reproduce.

I worry if we didn't do this, we wouldn't finish this any time soon, so hopefully it's fine.

MisRob commented:
> So, should I close this PR and raise new PRs with fewer files changed

Yes @Abhishek-Punhani, let's do it right now. Just choose maximum of 3 components. I'm sorry we will lose some of that other work, but you can keep it in a separate branch for future and gradually pull out from there into new very small PRs.

In a new PR, you can also include that list with places to test and then we can all have a look at it one more time and figure out better for each other steps to reproduce. Having that list will also help us to understand if we are actually looking into the same places.

I worry if we didn't do this, we wouldn't finish this any time soon, so hopefully it's fine.

MisRob commented:
> So, should I close this PR and raise new PRs with fewer files changed

Yes @Abhishek-Punhani, let's do it right now. Just choose maximum of 3 components. I'm sorry we will lose some of that other work, but you can keep it in a separate branch for future and gradually pull out from there into new very small PRs, so in a way nothing will be lost :) We'll just organize ourselves differently. I worry if we didn't do this, we wouldn't finish this any time soon, so hopefully it's fine. It's not meant to dismiss this work, but to help you. I haven't estimated the difficulty and now getting a better sense of this work.

In a new PR, you can also include that list with places (URLs or brief steps to navigate) to test and then we can all have a look at it one more time and figure out better for each other steps to reproduce. Having that list will help us to understand if we are actually looking into the same places, and @AllanOXDi and me will be more specific too. Kolibri re-uses many components at similar places and I think we're running into issues related to that.

Abhishek-Punhani commented:
Okay, @MisRob , As discussed, I will close this PR and raise new PRs with at most 3 components changed and will include a list of affected areas as well!

MisRob commented:
Thanks @Abhishek-Punhani
